gh-127081: use `getlogin_r` if available

duaneg opened the pull request:
The getlogin function is not thread-safe: replace with getlogin_r where available.

Note that this function is untested (unit test is skipped with a note it caused CI failures as behaviour differs between NIX environments).

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-127081 -->
* Issue: gh-127081
<!-- /gh-issue-number -->

duaneg committed:
gh-127081: use getlogin_r if available

The getlogin function is not thread-safe: replace with getlogin_r where
available.

Note that this function is untested (unit test is skipped with a note it caused
CI failures as behaviour differs between NIX environments).

colesbury commented on Modules/posixmodule.c:
```diff
@@ -9538,6 +9538,18 @@ os_getlogin_impl(PyObject *module)
     else
         result = PyErr_SetFromWindowsErr(GetLastError());
+#elif defined (HAVE_GETLOGIN_R)
+    /* AFAIK the maximum length should be 32, but this is not checkable */
```
The maximum username length is 32 [on Linux](https://man7.org/linux/man-pages/man5/utmp.5.html#:~:text=UT_LINESIZE%20%20%20%20%20%2032%0A%20%20%20%20%20%20%20%20%20%20%20%23define-,UT_NAMESIZE,-32%0A%20%20%20%20%20%20%20%20%20%20%20%23define%20UT_HOSTSIZE), but you also need an extra character for the null terminator.

https://github.com/bminor/glibc/blob/d58ebf51bca1a30e1a07e800e3823f4c96029da0/sysdeps/unix/getlogin_r.c#L82

It's 255 on macOS:
https://github.com/apple-oss-distributions/xnu/blob/8d741a5de7ff4191bf97d57b9f54c2f6d4a15585/bsd/sys/param.h#L97

duaneg replied:
Ah, thanks!

I was looking for a constant we could use, and those links are helpful. Unfortunately it seems Linux and other *NIX systems use different ones: `UT_NAMESIZE` on Linux and `MAXLOGNAME` on the BSDs and various others. I'll update `configure` to try and detect those and fall-back to a hard-coded value of 256 if both fail.

colesbury commented on Modules/posixmodule.c:
```diff
@@ -9538,6 +9538,18 @@ os_getlogin_impl(PyObject *module)
+        errno = -err;
+        posix_error();
+        errno = old_errno;
+    } else {
```
CPython formatting:

```suggestion
    }
    else {
```

duaneg committed:
Check for MAXLOGNAME and UT_NAMESIZE in configure and use them (in that order)
if available, falling back to a hard-coded maximum login name length of 255 if
neither is available.

duaneg committed:
Fix-up pyconfig.h.in

ZeroIntensity reviewed (commented):
This should get a test case.

colesbury reviewed (approved):
This LGTM assuming the build bots pass
